pkg/rosa/externalauthproviders.go

```diff
@@ -0,0 +1,136 @@
+type ExternalAuthClient struct {
+	ocm *sdk.Connection
+}
+
+// NewExternalAuthClient creates and return a new client to handle externalAuth operations.
+func NewExternalAuthClient(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (*ExternalAuthClient, error) {
+	ocmConnection, err := newOCMRawConnection(ctx, rosaScope)
+	if err != nil {
+		return nil, err
+	}
+	return &ExternalAuthClient{
+		ocm: ocmConnection,
+	}, nil
+}
+
+// Close closes the underlying ocm connection.
+func (c *ExternalAuthClient) Close() error {
+	return c.ocm.Close()
+}
+
+// CreateExternalAuth creates a new external auth porivder.
+func (c *ExternalAuthClient) CreateExternalAuth(clusterID string, externalAuth *cmv1.ExternalAuth) (*cmv1.ExternalAuth, error) {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).
+		ExternalAuthConfig().ExternalAuths().Add().Body(externalAuth).Send()
+	if err != nil {
+		return nil, handleErr(response.Error(), err)
+	}
+	return response.Body(), nil
+}
+
+// UpdateExternalAuth updates an existing external auth porivder.
+func (c *ExternalAuthClient) UpdateExternalAuth(clusterID string, externalAuth *cmv1.ExternalAuth) (*cmv1.ExternalAuth, error) {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).
+		ExternalAuthConfig().ExternalAuths().
+		ExternalAuth(externalAuth.ID()).
+		Update().Body(externalAuth).Send()
+	if err != nil {
+		return nil, handleErr(response.Error(), err)
+	}
+	return response.Body(), nil
+}
+
+// GetExternalAuth retrieves the specified external auth porivder.
+func (c *ExternalAuthClient) GetExternalAuth(clusterID string, externalAuthID string) (*cmv1.ExternalAuth, bool, error) {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).ExternalAuthConfig().
+		ExternalAuths().ExternalAuth(externalAuthID).
+		Get().
+		Send()
+	if response.Status() == 404 {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, handleErr(response.Error(), err)
+	}
+	return response.Body(), true, nil
+}
+
+// ListExternalAuths lists all external auth porivder for the cluster.
+func (c *ExternalAuthClient) ListExternalAuths(clusterID string) ([]*cmv1.ExternalAuth, error) {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).
+		ExternalAuthConfig().
+		ExternalAuths().
+		List().Page(1).Size(-1).
+		Send()
+	if err != nil {
+		return nil, handleErr(response.Error(), err)
+	}
+	return response.Items().Slice(), nil
+}
+
+// DeleteExternalAuth deletes the specified external auth porivder.
+func (c *ExternalAuthClient) DeleteExternalAuth(clusterID string, externalAuthID string) error {
+	response, err := c.ocm.ClustersMgmt().V1().
+		Clusters().Cluster(clusterID).
+		ExternalAuthConfig().ExternalAuths().
+		ExternalAuth(externalAuthID).
+		Delete().
+		Send()
+	if err != nil {
+		return handleErr(response.Error(), err)
+	}
+	return nil
+}
+
+// CreateBreakGlassCredential creates a break glass credential.
```

Comment:
What is a `break glass credential`? I'm assuming it's one that's used in case the OIDC isn't available?

Reply:
it can be used for that yes, but its main use case is to setup RBAC for oidc users. without RBAC, oidc users can login to the cluster but they wouldn't have any permissions.

Reply:
Could you please document that use in the comments? It's not clear from the usage here that its for RBAC.

Reply:
This is not resolved.

Reply:
n.b. also needs to be documented that we expect it to be non-functional for the majority of its life

Reply:
documented under `reconcileExternalAuthBootstrapKubeconfig()` func here https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4915/files#diff-693a964b12104ea3c2432ded8cbb95b43bf89782beb8d202e32dcf6422413e7cR596-R598
and here:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4915/files#diff-4e35dfebb4ba922251575193c9b544ad4c7e5c10fa1d118534474bd1c38dbc35R183-R186

Reply:
For the users. Having a kubeconfig for the guest on the CAPI cluster that is not functional is a surprising thing.

Reply:
I don't have strong feelings. My intuition is that CAPI maintainers would prefer to have it go in together, especially since it's a fairly straightforward ask.

Reply:
a dedicated doc section is planed to be added for the complete feature which will include all the details on usage and limitations.